patches/0003-tcp_proxy-Add-filter-state-proxy_read_before_connect.patch

```diff
@@ -243,28 +251,11 @@ index 8445d2f762..6ebeab02b2 100644
 +        StreamInfo::FilterState::StateType::ReadOnly,
 +        StreamInfo::FilterState::LifeSpan::Connection);
    }
- 
+
    const std::string key_;
-@@ -1353,14 +1365,25 @@ TEST_P(TcpProxyDynamicMetadataMatchIntegrationTest, DynamicMetadataMatch) {
-   initialize();
- 
-   expectEndpointToMatchRoute([](IntegrationTcpClient& tcp_client) -> std::string {
--    // Break the write into two; validate that the first is received before sending the second. This
--    // validates that a downstream filter can use this functionality, even if it can't make a
--    // decision after the first `onData()`.
--    EXPECT_TRUE(tcp_client.write("p", false));
--    tcp_client.waitForData("p");
-+    // Break the write into multiple chunks; validate that the first word is received before sending
-+    // the rest. This validates that a downstream filter can use this functionality, even if it
-+    // can't make a decision after the first `onData()`.
-+    EXPECT_TRUE(tcp_client.write("pri", false));
-+    // Writing additional data in multiple chunks to show that buffering of early data in tcp_buffer
-+    // works properly
-+    EXPECT_TRUE(tcp_client.write("mary is ", false));
-+    EXPECT_TRUE(tcp_client.write("selected ", false));
-+    EXPECT_TRUE(tcp_client.write("before ", false));
-+    // check that the 1st word is returned
-+    tcp_client.waitForData("primary");
```

Comment:
This change still should not be made.

Reply:
ah i missed this as previous convo was resolved, let me make a small update.

Reply:
The test is passed now, just a note that i removed the changes in our patch, as it seems to be conflicting with recent upstream changes.

```
vagrant@ubuntu-jammy:~/proxy$ bazel test @envoy//test/integration:tcp_proxy_integration_test
INFO: Analyzed target @envoy//test/integration:tcp_proxy_integration_test (309 packages loaded, 5520 targets configured).
INFO: Found 1 test target...
Target @envoy//test/integration:tcp_proxy_integration_test up-to-date:
  bazel-bin/external/envoy/test/integration/tcp_proxy_integration_test
INFO: Elapsed time: 10.413s, Critical Path: 0.23s
INFO: 1 process: 1 internal.
INFO: Build completed successfully, 1 total action

Executed 0 out of 1 test: 1 test passes.
```

Reply:
I have the test passing without any real changes from the previous patch:
```
diff --git a/patches/0003-tcp_proxy-Add-filter-state-proxy_read_before_connect.patch b/patches/0003-tcp_proxy-Add-filter-state-proxy_read_before_connect.patch
index 5561fe3e1..0c024beec 100644
--- a/patches/0003-tcp_proxy-Add-filter-state-proxy_read_before_connect.patch
+++ b/patches/0003-tcp_proxy-Add-filter-state-proxy_read_before_connect.patch
@@ -1,7 +1,7 @@
-From 736655de694fc2100ec9a752a4e2136ceb5339e9 Mon Sep 17 00:00:00 2001
+From e8369043306edba8921ab5ef71efd0522ff325c7 Mon Sep 17 00:00:00 2001
 From: Jarno Rajahalme <jarno@isovalent.com>
-Date: Tue, 27 Sep 2022 15:51:46 +0300
-Subject: [PATCH 3/4] tcp_proxy: Add filter state proxy_read_before_connect
+Date: Mon, 8 Apr 2024 15:21:08 +0200
+Subject: [PATCH 3/6] tcp_proxy: Add filter state proxy_read_before_connect
 
 Add filter state object bool proxy_receive_before_connect that tcp_proxy
 filter checks for and when set to 'true' will not set readDisable on the
@@ -29,6 +29,13 @@ reliant on the tcp_proxy internal detail, such as balancing and timing of
 the readDisable() calls on the downstream connection.
 
 Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>
+---
+ source/common/tcp_proxy/BUILD                 |  1 +
+ source/common/tcp_proxy/tcp_proxy.cc          | 34 ++++++++--
+ source/common/tcp_proxy/tcp_proxy.h           |  6 ++
+ test/integration/BUILD                        |  1 +
+ .../integration/tcp_proxy_integration_test.cc | 68 +++++++++++++------
+ 5 files changed, 84 insertions(+), 26 deletions(-)
 
 diff --git a/source/common/tcp_proxy/BUILD b/source/common/tcp_proxy/BUILD
 index 07d5c5995d..75f9047969 100644
@@ -43,7 +50,7 @@ index 07d5c5995d..75f9047969 100644
          "//envoy/tcp:conn_pool_interface",
          "//envoy/tcp:upstream_interface",
 diff --git a/source/common/tcp_proxy/tcp_proxy.cc b/source/common/tcp_proxy/tcp_proxy.cc
-index 16afd76f5b..858f15d9f6 100644
+index ee96e1b568..f11c797c21 100644
 --- a/source/common/tcp_proxy/tcp_proxy.cc
 +++ b/source/common/tcp_proxy/tcp_proxy.cc
 @@ -13,6 +13,7 @@
@@ -54,7 +61,7 @@ index 16afd76f5b..858f15d9f6 100644
  #include "envoy/upstream/cluster_manager.h"
  #include "envoy/upstream/upstream.h"
  
-@@ -256,7 +257,17 @@ void Filter::initialize(Network::ReadFilterCallbacks& callbacks, bool set_connec
+@@ -253,7 +254,17 @@ void Filter::initialize(Network::ReadFilterCallbacks& callbacks, bool set_connec
    // Need to disable reads so that we don't write to an upstream that might fail
    // in onData(). This will get re-enabled when the upstream connection is
    // established.
@@ -73,7 +80,7 @@ index 16afd76f5b..858f15d9f6 100644
    getStreamInfo().setDownstreamBytesMeter(std::make_shared<StreamInfo::BytesMeter>());
    getStreamInfo().setUpstreamInfo(std::make_shared<StreamInfo::UpstreamInfoImpl>());
  
-@@ -479,8 +490,11 @@ Network::FilterStatus Filter::establishUpstreamConnection() {
+@@ -489,8 +500,11 @@ Network::FilterStatus Filter::establishUpstreamConnection() {
      // cluster->trafficStats()->upstream_cx_none_healthy in the latter case.
      getStreamInfo().setResponseFlag(StreamInfo::ResponseFlag::NoHealthyUpstream);
      onInitFailure(UpstreamFailureReason::NoHealthyUpstream);
@@ -86,7 +93,7 @@ index 16afd76f5b..858f15d9f6 100644
  }
  
  void Filter::onClusterDiscoveryCompletion(Upstream::ClusterDiscoveryStatus cluster_status) {
-@@ -692,12 +706,18 @@ Network::FilterStatus Filter::onData(Buffer::Instance& data, bool end_stream) {
+@@ -705,12 +719,18 @@ Network::FilterStatus Filter::onData(Buffer::Instance& data, bool end_stream) {
    if (upstream_) {
      getStreamInfo().getUpstreamBytesMeter()->addWireBytesSent(data.length());
      upstream_->encodeData(data, end_stream);
@@ -106,7 +113,7 @@ index 16afd76f5b..858f15d9f6 100644
    return Network::FilterStatus::StopIteration;
  }
  
-@@ -814,7 +834,13 @@ void Filter::onUpstreamConnection() {
+@@ -827,7 +847,13 @@ void Filter::onUpstreamConnection() {
    connecting_ = false;
    // Re-enable downstream reads now that the upstream connection is established
    // so we have a place to send downstream data to.
@@ -122,7 +129,7 @@ index 16afd76f5b..858f15d9f6 100644
    read_callbacks_->upstreamHost()->outlierDetector().putResult(
        Upstream::Outlier::Result::LocalOriginConnectSuccessFinal);
 diff --git a/source/common/tcp_proxy/tcp_proxy.h b/source/common/tcp_proxy/tcp_proxy.h
-index 41697d2f0d..41f02eaf51 100644
+index 963a4f16bb..96af4c50a5 100644
 --- a/source/common/tcp_proxy/tcp_proxy.h
 +++ b/source/common/tcp_proxy/tcp_proxy.h
 @@ -22,6 +22,7 @@
@@ -142,7 +149,7 @@ index 41697d2f0d..41f02eaf51 100644
  /**
   * All tcp proxy stats. @see stats_macros.h
   */
-@@ -546,6 +549,9 @@ protected:
+@@ -551,6 +554,9 @@ protected:
    uint32_t connect_attempts_{};
    bool connecting_{};
    bool downstream_closed_{};
@@ -153,10 +160,10 @@ index 41697d2f0d..41f02eaf51 100644
  
  // This class deals with an upstream connection that needs to finish flushing, when the downstream
 diff --git a/test/integration/BUILD b/test/integration/BUILD
-index d666674781..08eee106e2 100644
+index 51937089c3..80f92c77ed 100644
 --- a/test/integration/BUILD
 +++ b/test/integration/BUILD
-@@ -1757,6 +1757,7 @@ envoy_cc_test(
+@@ -1785,6 +1785,7 @@ envoy_cc_test(
          "//source/common/event:dispatcher_includes",
          "//source/common/event:dispatcher_lib",
          "//source/common/network:utility_lib",
@@ -165,7 +172,7 @@ index d666674781..08eee106e2 100644
          "//source/extensions/filters/network/common:factory_base_lib",
          "//source/extensions/filters/network/tcp_proxy:config",
 diff --git a/test/integration/tcp_proxy_integration_test.cc b/test/integration/tcp_proxy_integration_test.cc
-index 8445d2f762..6ebeab02b2 100644
+index 90ad072b81..95a6b1cf80 100644
 --- a/test/integration/tcp_proxy_integration_test.cc
 +++ b/test/integration/tcp_proxy_integration_test.cc
 @@ -12,6 +12,8 @@
@@ -177,7 +184,7 @@ index 8445d2f762..6ebeab02b2 100644
  #include "source/extensions/filters/network/common/factory_base.h"
  #include "source/extensions/transport_sockets/tls/context_manager_impl.h"
  
-@@ -1258,18 +1260,27 @@ public:
+@@ -1299,18 +1301,27 @@ public:
    Network::FilterStatus onData(Buffer::Instance& data, bool) override {
      if (!metadata_set_) {
        // To allow testing a write that returns `StopIteration`, only proceed
@@ -213,7 +220,7 @@ index 8445d2f762..6ebeab02b2 100644
  
        ProtobufWkt::Struct& map =
            (*read_callbacks_->connection()
-@@ -1278,24 +1289,25 @@ public:
+@@ -1319,24 +1330,25 @@ public:
                  .mutable_filter_metadata())[Envoy::Config::MetadataFilters::get().ENVOY_LB];
        (*map.mutable_fields())[key_] = val;
  
@@ -245,13 +252,13 @@ index 8445d2f762..6ebeab02b2 100644
    }
  
    const std::string key_;
-@@ -1353,14 +1365,25 @@ TEST_P(TcpProxyDynamicMetadataMatchIntegrationTest, DynamicMetadataMatch) {
+@@ -1394,14 +1406,26 @@ TEST_P(TcpProxyDynamicMetadataMatchIntegrationTest, DynamicMetadataMatch) {
    initialize();
  
    expectEndpointToMatchRoute([](IntegrationTcpClient& tcp_client) -> std::string {
 -    // Break the write into two; validate that the first is received before sending the second. This
--    // validates that a downstream filter can use this functionality, even if it can't make a
--    // decision after the first `onData()`.
+-    // validates that a downstream network filter can use this functionality, even if it can't make
+-    // a decision after the first `onData()`.
 -    EXPECT_TRUE(tcp_client.write("p", false));
 -    tcp_client.waitForData("p");
 +    // Break the write into multiple chunks; validate that the first word is received before sending
@@ -265,6 +272,7 @@ index 8445d2f762..6ebeab02b2 100644
 +    EXPECT_TRUE(tcp_client.write("before ", false));
 +    // check that the 1st word is returned
 +    tcp_client.waitForData("primary");
++
      tcp_client.clearData();
 -    EXPECT_TRUE(tcp_client.write("rimary", false));
 -    return "primary";
@@ -278,7 +286,7 @@ index 8445d2f762..6ebeab02b2 100644
    });
  }
  
-@@ -1377,7 +1400,7 @@ TEST_P(TcpProxyDynamicMetadataMatchIntegrationTest, DynamicMetadataNonMatch) {
+@@ -1418,7 +1442,7 @@ TEST_P(TcpProxyDynamicMetadataMatchIntegrationTest, DynamicMetadataNonMatch) {
  
    initialize();
  
@@ -288,5 +296,5 @@ index 8445d2f762..6ebeab02b2 100644
  
  INSTANTIATE_TEST_SUITE_P(TcpProxyIntegrationTestParams, TcpProxySslIntegrationTest,
 -- 
-2.41.0
+2.44.0
 
```

test:
```
$ bazel test @envoy//test/integration:tcp_proxy_integration_test
INFO: Analyzed target @envoy//test/integration:tcp_proxy_integration_test (282 packages loaded, 4504 targets configured).
INFO: Found 1 test target...
Target @envoy//test/integration:tcp_proxy_integration_test up-to-date:
  bazel-bin/external/envoy/test/integration/tcp_proxy_integration_test
INFO: Elapsed time: 60.197s, Critical Path: 51.47s
INFO: 4 processes: 1 internal, 3 linux-sandbox.
INFO: Build completed successfully, 4 total actions

Executed 1 out of 1 test: 1 test passes.
```

Reply:
Somehow I messed up the previous patch, and locally it got stuck with recent changes related access logger, then the test is timed out.

Let's use your patch, thanks a lot.